wallet/manager.go

```diff
@@ -265,6 +272,240 @@ func (m *Manager) Reserve(ids []types.Hash256, duration time.Duration) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	knownAddresses := make(map[types.Address]bool)
+	relevantAddr := func(addr types.Address) bool {
+		if exists, ok := knownAddresses[addr]; ok {
+			return exists
+		}
+		_, err := m.store.WalletAddress(walletID, addr)
+		if errors.Is(err, ErrNotFound) {
+			knownAddresses[addr] = false
+			return false
+		} else if err != nil {
+			panic(err)
+		}
+		knownAddresses[addr] = true
+		return true
+	}
+
+	ephemeral := make(map[types.SiacoinOutputID]types.SiacoinElement)
+	inPool := make(map[types.SiacoinOutputID]bool)
+	for _, txn := range m.chain.PoolTransactions() {
+		for _, sci := range txn.SiacoinInputs {
+			inPool[sci.ParentID] = true
+			delete(ephemeral, sci.ParentID)
+		}
+		for i, sco := range txn.SiacoinOutputs {
+			if relevantAddr(sco.Address) {
+				scoid := txn.SiacoinOutputID(i)
+				ephemeral[scoid] = types.SiacoinElement{
+					ID:            scoid,
+					StateElement:  types.StateElement{LeafIndex: types.UnassignedLeafIndex},
+					SiacoinOutput: sco,
+				}
+			}
+		}
+	}
+	for _, txn := range m.chain.V2PoolTransactions() {
+		for _, sci := range txn.SiacoinInputs {
+			inPool[sci.Parent.ID] = true
+			delete(ephemeral, sci.Parent.ID)
+		}
+		for i, sco := range txn.SiacoinOutputs {
+			if relevantAddr(sco.Address) {
+				sce := txn.EphemeralSiacoinOutput(i)
+				ephemeral[sce.ID] = sce
+			}
+		}
+	}
+
+	tip := m.chain.Tip()
+	if amount.IsZero() {
+		return nil, tip, types.ZeroCurrency, nil
+	}
+
+	var inputSum types.Currency
+	var selected []types.SiacoinElement
+	const utxoBatchSize = 100
+top:
+	for i := 0; ; i += utxoBatchSize {
+		// extra large wallets may need to paginate through utxos
+		// to find enough to cover the amount
+		utxos, err := m.store.WalletSiacoinOutputs(walletID, tip, i, utxoBatchSize)
+		if err != nil {
+			return nil, types.ChainIndex{}, types.ZeroCurrency, fmt.Errorf("failed to get siacoin elements: %w", err)
+		} else if len(utxos) == 0 {
+			return nil, types.ChainIndex{}, types.ZeroCurrency, ErrInsufficientFunds
+		}
+
+		for _, sce := range utxos {
+			if inPool[sce.ID] || m.used[types.Hash256(sce.ID)] {
+				continue
+			}
+
+			selected = append(selected, sce)
+			inputSum = inputSum.Add(sce.SiacoinOutput.Value)
+			if inputSum.Cmp(amount) >= 0 {
+				break top
+			}
+		}
+	}
+
+	if inputSum.Cmp(amount) < 0 {
+		if !useUnconfirmed {
+			return nil, types.ChainIndex{}, types.ZeroCurrency, ErrInsufficientFunds
+		}
+
+		for _, sce := range ephemeral {
+			if inPool[sce.ID] || m.used[types.Hash256(sce.ID)] {
+				continue
+			}
+
+			selected = append(selected, sce)
+			inputSum = inputSum.Add(sce.SiacoinOutput.Value)
+			if inputSum.Cmp(amount) >= 0 {
+				break
+			}
+		}
+	}
+
+	if inputSum.Cmp(amount) < 0 {
+		return nil, types.ChainIndex{}, types.ZeroCurrency, ErrInsufficientFunds
+	}
+
+	for _, sce := range selected {
+		m.used[types.Hash256(sce.ID)] = true
```

Comment:
Instead of using `AfterFunc` it would be a bit more efficient, but maybe less foolproof, to set this to the expiration time rather than `true`.

Reply:
If you set the expiration time, you need to loop through and prune the map occasionally. Locking and looping through all the locked elements might end up less efficient than a separate timer to delete specific items on larger nodes.

Reply:
There is an edge case now though where the elements may be released before the initial timer expires. So I'll change it.